book/src/dev/rfcs/0010-v5-transaction.md

```diff
@@ -0,0 +1,224 @@
+
+
+```
+V4 {
+    inputs: Vec<transparent::Input>,
+    outputs: Vec<transparent::Output>,
+    lock_time: LockTime,
+    expiry_height: block::Height,
+    value_balance: Amount,
+    joinsplit_data: Option<JoinSplitData<Groth16Proof>>,
+    shielded_data: Option<ShieldedData>,
+}
+```
+
+Currently the `ShieldedData` type is defined in `zebra-chain/src/transaction/shielded_data.rs` as follows:
+
+```
+pub struct ShieldedData {
+    pub first: Either<Spend, Output>,
+    pub rest_spends: Vec<Spend>,
+    pub rest_outputs: Vec<Output>,
+    pub binding_sig: Signature<Binding>,
+}
+```
+
+We know by protocol (2nd table of [Transaction Encoding and Consensus](https://zips.z.cash/protocol/nu5.pdf#txnencodingandconsensus)) that V5 transactions will support sapling data however we also know by protocol that spends ([Spend Description Encoding and Consensus](https://zips.z.cash/protocol/nu5.pdf#spendencodingandconsensus), See †) and outputs ([Output Description Encoding and Consensus](https://zips.z.cash/protocol/nu5.pdf#outputencodingandconsensus), See †) fields change from V4 to V5.
+
+Here we have the proposed changes for V4 transactions:
+
+```
+V4 {
+    ...
+    sapling_value_balance: Amount,
+    ...
+    sapling_shielded_data: Option<SaplingShieldedData::V4>,
+}
+```
+
+In order to mantain the order of the fields that we will need for Serialization/Deserialization we keep the `value_balance` in the transaction for V4 but we rename it to `sapling_value_balance` for clarity.
+
+`ShieldedData` is currently defined and implemented in `src/transaction/shielded_data.rs`. We propose to rename this file to `src/transaction/sapling_shielded_data.rs`. Inside we will change `ShieldedData` into a `SaplingShieldedData` enum with `V4` and `V5` variants.
+
+```
+pub enum SaplingShieldedData {
+    /// Either a spend or output description.
+    ///
+    /// Storing this separately ensures that it is impossible to construct
+    /// an invalid `SaplingShieldedData` with no spends or outputs.
+    /// ...
+    V4 {
+        pub first: Either<SaplingSpend::V4, SaplingOutput>,
+        pub rest_spends: Vec<SaplingSpend::V4>,
+        pub rest_outputs: Vec<SaplingOutput>,
+        pub binding_sig: Signature<Binding>,
+    },
+    V5 {
+        pub first: Either<SaplingSpend::V5, SaplingOutput>,
+        pub rest_spends: Vec<SaplingSpend::V5>,
+        pub rest_outputs: Vec<SaplingOutput>,
+        pub sapling_value_balance: Amount,
+        pub anchor: tree::Root,
+        pub binding_sig: Signature<Binding>,
+    }
+}
+```
+
+Proposed `SaplingSpend` is now defined as an enum with `V4` and `V5` variants. Sapling spend code is located at `zebra-chain/src/sapling/spend.rs`. Notable difference here is that the `anchor` in `V4` is needed for every spend of the transaction while in `V5` the anchor is a single one defined in `SaplingShieldedData`:
+
+```
+pub enum SaplingSpend {
```

Comment:
Since Spend descriptions are specific to Sapling and it will live in the `sapling` module, I think we can safely keep this name as `Spend`:
```suggestion
pub enum Spend {
```

Reply:
Yep, this was my suggestion, but yours is better.

We can just use `sapling::Spend` outside the module, like we use `block::Hash` and `block::Height`.